prediction_market_agent/agents/safe_guard_agent/guards/llm.py

```diff
@@ -23,49 +27,61 @@
 HISTORY_LIMIT = 25
 
 
-@observe()
-def validate_safe_transaction_llm(
-    new_transaction: DetailedTransactionResponse,
-    new_transaction_safetx: SafeTx,
-    all_addresses_from_tx: list[ChecksumAddress],
-    history: list[DetailedTransactionResponse],
-) -> ValidationResult:
-    # Get the latest ones.
-    history = sorted(
-        history, key=lambda x: x.executedAt or float("-inf"), reverse=True
-    )[:HISTORY_LIMIT]
-
-    agent = Agent(
-        OpenAIModel(
-            "gpt-4o",
-            provider=get_openai_provider(api_key=APIKeys().openai_api_key),
-        ),
-        system_prompt="""You are fraud detection agent. 
+class LLMValidationResult(BaseModel):
+    reason: str
+    ok: bool
+
+
+class LLM(AbstractGuard):
+    name = "LLM"
+    description = "This guard uses a large language model to analyze the transaction and determine if it is malicious."
+
+    @observe(name="validate_safe_transaction_llm")
+    def validate(
+        self,
+        new_transaction: DetailedTransactionResponse,
+        new_transaction_safetx: SafeTx,
+        all_addresses_from_tx: list[ChecksumAddress],
+        history: list[DetailedTransactionResponse],
+    ) -> ValidationResult:
+        # Get the latest ones.
+        history = sorted(
+            history, key=lambda x: x.executedAt or float("-inf"), reverse=True
+        )[:HISTORY_LIMIT]
+
+        agent = Agent(
+            OpenAIModel(
+                "gpt-4o",
```

Comment:
I think I asked before, but - did you try using a cheaper model? Maybe this can be investigated in the evals, if gpt-4o-mini is too bad, otherwise we could use that (or even make the model a dropdown choice on the app)

Reply:
Previously it was `o1`, so this is the cheaper one 😄 I wouldn't go as cheap as possible here. Traffic is low, and it needs to be at least somewhat clever. This has ~25k input tokens and ~100 output tokens, that's ~$0.127 per run.

Agree with evals, but I'm afraid we will need more expensive models, not cheaper.